File: test/unit/roon-proto.test.js
"use strict";
/*
 * The two wire formats, tested without a Roon Core.
 *
 * There is no Core in CI and there never will be, so this is the only part of
 * the Roon support that can be checked completely — which is why the codecs
 * were separated out in the first place. The failure these guard against is
 * not a crash: a u16 read little-endian, or a body taken one byte short, makes
 * a Core look absent or a library look empty, and "Roon found nothing" is
 * indistinguishable from "Roon is not running".
 */
const test = require("node:test");
const assert = require("node:assert");
const P = require("../../lib/roon-proto");

// ------------------------------------------------------------------- SOOD

/** Build a reply packet the way a Core does, so the parser can be checked. */
function soodReply(props) {
  const parts = [Buffer.from("SOOD", "utf8"), Buffer.from([2, "R".charCodeAt(0)])];
  for (const [name, value] of Object.entries(props)) {
    const n = Buffer.from(name, "utf8");
    const head = Buffer.alloc(1 + n.length + 2);
    head.writeUInt8(n.length, 0);
    n.copy(head, 1);
    if (value === null) {
      head.writeUInt16BE(0xffff, 1 + n.length);
      parts.push(head);
    } else {
      const v = Buffer.from(String(value), "utf8");
      head.writeUInt16BE(v.length, 1 + n.length);
      parts.push(head, v);
    }
  }
  return Buffer.concat(parts);
}

test("a SOOD query names the Roon service id and nothing else", () => {
  const q = P.buildSoodQuery("abc");
  assert.strictEqual(q.toString("utf8", 0, 4), "SOOD");
  assert.strictEqual(q[4], 2, "protocol version");
  assert.strictEqual(String.fromCharCode(q[5]), "Q");

  // Decoded, not merely searched for. A query whose length fields are written
  // the wrong way round still CONTAINS the service id, and is still answered
  // by no Core whatsoever — which reads as "there is no Roon on this network"
  // and sends the user looking at their Wi-Fi.
  const packet = P.parseSoodPacket(q);
  assert.strictEqual(packet.kind, "Q");
  assert.deepStrictEqual(Object.keys(packet.props).sort(),
    ["_tid", "query_service_id"]);
  assert.strictEqual(packet.props.query_service_id, P.ROON_SERVICE_ID);
  assert.strictEqual(packet.props._tid, "abc");
});

test("a long value in a query survives its own length field", () => {
  // 300 bytes again, from the writing side this time: 0x012C little-endian is
  // 0x2C01, and the property walk then runs off the end of the packet.
  const packet = P.parseSoodPacket(P.buildSoodQuery("t".repeat(300)));
  assert.ok(packet, "the query must be parseable by the same walker");
  assert.strictEqual(packet.props._tid.length, 300);
});

test("a SOOD reply is parsed, including a null value", () => {
  const props = P.parseSoodReply(soodReply({
    unique_id: "1f2e", http_port: "9330", name: "Study", tid: null,
  }));
  assert.strictEqual(props.unique_id, "1f2e");
  assert.strictEqual(props.http_port, "9330");
  assert.strictEqual(props.name, "Study");
  assert.strictEqual(props.tid, null, "0xFFFF is null, not empty string");
});

test("an empty string value is not the same as a null one", () => {
  const props = P.parseSoodReply(soodReply({ unique_id: "x", http_port: "1", name: "" }));
  assert.strictEqual(props.name, "", "0 length is an empty value");
  assert.notStrictEqual(props.name, null);
});

test("a two-byte length is read big-endian", () => {
  // 300 bytes: 0x012C. Read the other way round it is 11265 and the parse
  // walks off the end, which is what a little-endian read looks like.
  const long = "n".repeat(300);
  const props = P.parseSoodReply(soodReply({ unique_id: "x", http_port: "1", name: long }));
  assert.strictEqual(props.name.length, 300);
});

test("packets that are not SOOD replies are ignored rather than thrown on", () => {
  assert.strictEqual(P.parseSoodReply(Buffer.from("hello world")), null);
  assert.strictEqual(P.parseSoodReply(Buffer.alloc(0)), null);
  assert.strictEqual(P.parseSoodReply(null), null);
  // Our own query, which arrives back on a broadcast socket.
  assert.strictEqual(P.parseSoodReply(P.buildSoodQuery("t")), null,
    "a query is not a reply");
  // Truncated mid-value.
  const good = soodReply({ unique_id: "abcdef", http_port: "9330" });
  assert.strictEqual(P.parseSoodReply(good.subarray(0, good.length - 3)), null);
});

test("a Core is only usable if it said where to connect", () => {
  const withAddr = P.soodCore({ unique_id: "u", http_port: "9330", _replyaddr: "10.0.0.5" });
  assert.deepStrictEqual(withAddr,
    { host: "10.0.0.5", port: 9330, uniqueId: "u", displayName: "" });

  // No _replyaddr: the packet's own source address is used.
  assert.strictEqual(P.soodCore({ unique_id: "u", http_port: "9330" }, "10.0.0.9").host,
    "10.0.0.9");

  assert.strictEqual(P.soodCore({ unique_id: "u" }, "10.0.0.9"), null,
    "no port means nowhere to connect");
  assert.strictEqual(P.soodCore({ http_port: "9330" }, "10.0.0.9"), null);
  assert.strictEqual(P.soodCore({ unique_id: "u", http_port: "0" }, "10.0.0.9"), null);
  assert.strictEqual(P.soodCore(null, "10.0.0.9"), null);
});

// -------------------------------------------------------------------- MOO

test("a MOO request round-trips", () => {
  const frame = P.encodeRequest("com.roonlabs.browse:1", "load", 7,
    { hierarchy: "albums", offset: 0, count: 100 });
  const msg = P.parseMoo(frame);
  assert.strictEqual(msg.verb, "REQUEST");
  assert.strictEqual(msg.service, "com.roonlabs.browse:1");
  assert.strictEqual(msg.name, "load");
  assert.strictEqual(msg.requestId, "7");
  assert.deepStrictEqual(P.mooJson(msg), { hierarchy: "albums", offset: 0, count: 100 });
});

test("a response carries a result name and no service", () => {
  const frame = P.encodeMoo("COMPLETE", "Registered", 1,
    Buffer.from('{"core_id":"c1","token":"t"}', "utf8"));
  const msg = P.parseMoo(frame);
  assert.strictEqual(msg.verb, "COMPLETE");
  assert.strictEqual(msg.service, null, "only a REQUEST names a service");
  assert.strictEqual(msg.name, "Registered");
  assert.strictEqual(P.mooJson(msg).token, "t");
});

test("a bodiless message parses, and its body is null rather than empty", () => {
  const msg = P.parseMoo(P.encodeMoo("COMPLETE", "Success", 3, null));
  assert.strictEqual(msg.name, "Success");
  assert.strictEqual(msg.body, null);
  assert.strictEqual(msg.bodyText, null);
  assert.strictEqual(P.mooJson(msg), null);
});

test("a body containing a blank line is read whole", () => {
  // The reason this parser counts bytes instead of splitting on "\n\n". An
  // album title with a newline in it would otherwise truncate the message and
  // the whole page of albums would be lost as unparseable.
  const body = Buffer.from(JSON.stringify({ title: "one\n\ntwo", n: 1 }), "utf8");
  const msg = P.parseMoo(P.encodeMoo("COMPLETE", "Success", 4, body));
  assert.deepStrictEqual(P.mooJson(msg), { title: "one\n\ntwo", n: 1 });
});

test("a body is measured in bytes, not characters", () => {
  // "é" is two bytes. Content-Length written as a character count would cut
  // the JSON short and every non-ASCII album title in a page would break it.
  const body = Buffer.from(JSON.stringify({ title: "Café Bleu — Émilie" }), "utf8");
  const frame = P.encodeMoo("COMPLETE", "Success", 5, body);
  assert.ok(frame.includes(Buffer.from("Content-Length: " + body.length)),
    "the header states the byte length");
  assert.strictEqual(P.mooJson(P.parseMoo(frame)).title, "Café Bleu — Émilie");
});

test("headers other than the two that matter are kept", () => {
  const frame = Buffer.from(
    "MOO/1 COMPLETE Success\nRequest-Id: 2\nLogging: async\n\n", "utf8");
  const msg = P.parseMoo(frame);
  assert.strictEqual(msg.headers.Logging, "async");
  assert.strictEqual(msg.requestId, "2");
});

test("a truncated frame is not a message", () => {
  const body = Buffer.from('{"a":1}', "utf8");
  const frame = P.encodeMoo("COMPLETE", "Success", 6, body);
  // Claims seven body bytes, carries four.
  assert.strictEqual(P.parseMoo(frame.subarray(0, frame.length - 3)), null,
    "a short body must not be handed on as a good message");
  // No blank line at all: still arriving.
  assert.strictEqual(P.parseMoo(Buffer.from("MOO/1 COMPLETE Success\n", "utf8")), null);
  // No Request-Id: nothing to correlate it with.
  assert.strictEqual(P.parseMoo(Buffer.from("MOO/1 COMPLETE Success\n\n", "utf8")), null);
});

test("something that is not MOO at all is ignored", () => {
  assert.strictEqual(P.parseMoo(Buffer.from("GET / HTTP/1.1\n\n", "utf8")), null);
  assert.strictEqual(P.parseMoo(Buffer.alloc(0)), null);
  assert.strictEqual(P.parseMoo(null), null);
  assert.strictEqual(P.parseMoo(Buffer.from("MOO/1 COMPLETE\n\n", "utf8")), null,
    "a header line with no result name");
});

test("a malformed JSON body is null, not a thrown parse error", () => {
  const msg = P.parseMoo(P.encodeMoo("COMPLETE", "Success", 8,
    Buffer.from("{not json", "utf8")));
  assert.strictEqual(msg.bodyText, "{not json");
  assert.strictEqual(P.mooJson(msg), null);
});

File: android/core/src/test/kotlin/com/musicd/migrate/roon/RoonProtoTest.kt
package com.musicd.migrate.roon

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.ByteArrayOutputStream

/**
 * The two wire formats, tested without a Roon Core.
 *
 * The twin of test/unit/roon-proto.test.js. There is no Core in CI and there
 * never will be, so this is the only part of the Roon support that can be
 * checked completely — which is why the codecs were separated out. The failure
 * these guard against is not a crash: a u16 read little-endian, or a body
 * taken one byte short, makes a Core look absent or a library look empty, and
 * "Roon found nothing" is indistinguishable from "Roon is not running".
 */
class RoonProtoTest {

    /** Build a reply packet the way a Core does, so the parser can be checked. */
    private fun soodReply(props: List<Pair<String, String?>>): ByteArray {
        val out = ByteArrayOutputStream()
        out.write("SOOD".toByteArray())
        out.write(2); out.write('R'.code)
        for ((name, value) in props) {
            val n = name.toByteArray(Charsets.UTF_8)
            out.write(n.size)
            out.write(n)
            if (value == null) {
                out.write(0xFF); out.write(0xFF)
            } else {
                val v = value.toByteArray(Charsets.UTF_8)
                out.write((v.size shr 8) and 0xFF)
                out.write(v.size and 0xFF)
                out.write(v)
            }
        }
        return out.toByteArray()
    }

    // ------------------------------------------------------------------ SOOD

    @Test fun `a SOOD query names the Roon service id and nothing else`() {
        val q = RoonProto.buildSoodQuery("abc")
        assertEquals("SOOD", String(q, 0, 4))
        assertEquals(2, q[4].toInt())
        assertEquals('Q', q[5].toInt().toChar())

        // Decoded, not merely searched for. A query whose length fields are
        // written the wrong way round still CONTAINS the service id, and is
        // still answered by no Core whatsoever — which reads as "there is no
        // Roon on this network" and sends the user looking at their Wi-Fi.
        val packet = RoonProto.parseSoodPacket(q)!!
        assertEquals('Q', packet.kind)
        assertEquals(setOf("_tid", "query_service_id"), packet.props.keys)
        assertEquals(RoonProto.ROON_SERVICE_ID, packet.props["query_service_id"])
        assertEquals("abc", packet.props["_tid"])
    }

    @Test fun `a long value in a query survives its own length field`() {
        // 300 bytes: 0x012C little-endian is 0x2C01, and the property walk
        // then runs off the end of the packet.
        val packet = RoonProto.parseSoodPacket(RoonProto.buildSoodQuery("t".repeat(300)))
        assertNotNull("the query must be parseable by the same walker", packet)
        assertEquals(300, packet!!.props["_tid"]!!.length)
    }

    @Test fun `a SOOD reply is parsed, including a null value`() {
        val props = RoonProto.parseSoodReply(soodReply(listOf(
            "unique_id" to "1f2e", "http_port" to "9330", "name" to "Study", "tid" to null)))!!
        assertEquals("1f2e", props["unique_id"])
        assertEquals("9330", props["http_port"])
        assertEquals("Study", props["name"])
        assertTrue("0xFFFF is null, not an empty string", props.containsKey("tid"))
        assertNull(props["tid"])
    }

    @Test fun `an empty string value is not the same as a null one`() {
        val props = RoonProto.parseSoodReply(soodReply(listOf(
            "unique_id" to "x", "http_port" to "1", "name" to "")))!!
        assertEquals("", props["name"])
    }

    @Test fun `a two-byte length is read big-endian`() {
        val long = "n".repeat(300)
        val props = RoonProto.parseSoodReply(soodReply(listOf(
            "unique_id" to "x", "http_port" to "1", "name" to long)))!!
        assertEquals(300, props["name"]!!.length)
    }

    @Test fun `packets that are not SOOD replies are ignored rather than thrown on`() {
        assertNull(RoonProto.parseSoodReply("hello world".toByteArray()))
        assertNull(RoonProto.parseSoodReply(ByteArray(0)))
        assertNull(RoonProto.parseSoodReply(null))
        assertNull("a query is not a reply",
            RoonProto.parseSoodReply(RoonProto.buildSoodQuery("t")))
        val good = soodReply(listOf("unique_id" to "abcdef", "http_port" to "9330"))
        assertNull(RoonProto.parseSoodReply(good.copyOfRange(0, good.size - 3)))
    }

    @Test fun `a Core is only usable if it said where to connect`() {
        val withAddr = RoonProto.soodCore(
            mapOf("unique_id" to "u", "http_port" to "9330", "_replyaddr" to "10.0.0.5"), null)
        assertEquals(RoonProto.FoundCore("10.0.0.5", 9330, "u", ""), withAddr)

        assertEquals("10.0.0.9", RoonProto.soodCore(
            mapOf("unique_id" to "u", "http_port" to "9330"), "10.0.0.9")!!.host)

        assertNull("no port means nowhere to connect",
            RoonProto.soodCore(mapOf("unique_id" to "u"), "10.0.0.9"))
        assertNull(RoonProto.soodCore(mapOf("http_port" to "9330"), "10.0.0.9"))
        assertNull(RoonProto.soodCore(mapOf("unique_id" to "u", "http_port" to "0"), "10.0.0.9"))
        assertNull(RoonProto.soodCore(null, "10.0.0.9"))
    }

    // ------------------------------------------------------------------- MOO

    @Test fun `a MOO request round-trips`() {
        val frame = RoonProto.encodeRequest("com.roonlabs.browse:1", "load", 7,
            """{"hierarchy":"albums","offset":0}""")
        val msg = RoonProto.parse(frame)!!
        assertEquals("REQUEST", msg.verb)
        assertEquals("com.roonlabs.browse:1", msg.service)
        assertEquals("load", msg.name)
        assertEquals("7", msg.requestId)
        assertEquals("""{"hierarchy":"albums","offset":0}""", msg.bodyText)
    }

    @Test fun `a response carries a result name and no service`() {
        val frame = RoonProto.encode("COMPLETE", "Registered", 1,
            """{"core_id":"c1","token":"t"}""".toByteArray())
        val msg = RoonProto.parse(frame)!!
        assertEquals("COMPLETE", msg.verb)
        assertNull("only a REQUEST names a service", msg.service)
        assertEquals("Registered", msg.name)
    }

    @Test fun `a bodiless message parses, and its body is null rather than empty`() {
        val msg = RoonProto.parse(RoonProto.encode("COMPLETE", "Success", 3, null))!!
        assertEquals("Success", msg.name)
        assertNull(msg.body)
        assertNull(msg.bodyText)
    }

    @Test fun `a body containing a blank line is read whole`() {
        // The reason this parser counts bytes instead of splitting on "\n\n".
        // An album title with a newline in it would otherwise truncate the
        // message and the whole page of albums would be lost as unparseable.
        val body = "{\"title\":\"one\n\ntwo\"}".toByteArray(Charsets.UTF_8)
        val msg = RoonProto.parse(RoonProto.encode("COMPLETE", "Success", 4, body))!!
        assertEquals("{\"title\":\"one\n\ntwo\"}", msg.bodyText)
    }

    @Test fun `a body is measured in bytes, not characters`() {
        // "é" is two bytes. A Content-Length written as a character count
        // would cut the JSON short, and every non-ASCII album title in a page
        // would break it.
        val text = """{"title":"Café Bleu — Émilie"}"""
        val body = text.toByteArray(Charsets.UTF_8)
        val frame = RoonProto.encode("COMPLETE", "Success", 5, body)
        assertTrue(String(frame, Charsets.UTF_8).contains("Content-Length: ${body.size}"))
        assertEquals(text, RoonProto.parse(frame)!!.bodyText)
    }

    @Test fun `headers other than the two that matter are kept`() {
        val msg = RoonProto.parse(
            "MOO/1 COMPLETE Success\nRequest-Id: 2\nLogging: async\n\n".toByteArray())!!
        assertEquals("async", msg.headers["Logging"])
        assertEquals("2", msg.requestId)
    }

    @Test fun `a truncated frame is not a message`() {
        val frame = RoonProto.encode("COMPLETE", "Success", 6, """{"a":1}""".toByteArray())
        assertNull("a short body must not be handed on as a good message",
            RoonProto.parse(frame.copyOfRange(0, frame.size - 3)))
        assertNull(RoonProto.parse("MOO/1 COMPLETE Success\n".toByteArray()))
        assertNull("no Request-Id: nothing to correlate it with",
            RoonProto.parse("MOO/1 COMPLETE Success\n\n".toByteArray()))
    }

    @Test fun `something that is not MOO at all is ignored`() {
        assertNull(RoonProto.parse("GET / HTTP/1.1\n\n".toByteArray()))
        assertNull(RoonProto.parse(ByteArray(0)))
        assertNull(RoonProto.parse(null))
        assertNull(RoonProto.parse("MOO/1 COMPLETE\n\n".toByteArray()))
    }

    /**
     * The two implementations produce the same bytes.
     *
     * Not a formality: index.js and the APK talk to the same Core with the
     * same protocol, and a difference here is a feature that works in Docker
     * and silently does not on a phone. The JavaScript side's own fixtures are
     * in test/unit/roon-proto.test.js; these are the bytes it produces.
     */
    @Test fun `the SOOD query bytes match the JavaScript ones`() {
        // Pinned as hex, against what lib/roon-proto.js produces for the same
        // transaction id. Two implementations of one packet format, and a
        // Core answers neither if they disagree.
        assertEquals(
            "534f4f440251045f74696400036162631071756572795f73" +
            "6572766963655f6964002430303732303732342d35313433" +
            "2d346139622d616261632d306535306362613637346262",
            RoonProto.buildSoodQuery("abc").joinToString("") { "%02x".format(it) })
    }

    @Test fun `the frames match the ones lib roon-proto js produces`() {
        assertEquals(
            "MOO/1 REQUEST com.roonlabs.registry:1/info\nRequest-Id: 0\n\n",
            String(RoonProto.encodeRequest("com.roonlabs.registry:1", "info", 0, null),
                   Charsets.UTF_8))
        assertEquals(
            "MOO/1 COMPLETE Success\nRequest-Id: 4\nContent-Length: 7\n" +
                "Content-Type: application/json\n\n{\"a\":1}",
            String(RoonProto.encode("COMPLETE", "Success", 4, """{"a":1}""".toByteArray()),
                   Charsets.UTF_8))
    }
}
